### PR TITLE
Basic docs

### PR DESCRIPTION
**Pull request by kivimango**
I added some documentation to the basic stuff (widget macro, Template, and State traits).
I hope it's technically correct, please check @FloVanGH.
@tzargoth please check the grammar and spelling.
Thank you.

### CODE REVIEW

**Comment by FloVanGH** [comment]
Thank you, that is a good think. Will review it tomorrow.

**Comment by FloVanGH** [comment]
From my perspective there are now technically errors. And it's really good quality of documentation. Thank you.

Now we need also the input of @tzargoth.

File: crates/api/src/widget/state.rs [hunk 1]
```diff
@@ -7,23 +7,80 @@ pub trait AsAny: Any {
     fn as_any_mut(&mut self) -> &mut dyn Any;
 }
 
-/// Used to define a state of a widget.
+/// Used to define a state of a [`widget`].
+/// 
+/// The state holds the logic of a widget which makes it interactive.
+/// The state of a widget made up of a struct which implements this trait with its fields and its methods.
```
Comment (tzargoth) [inline crates/api/src/widget/state.rs:13]: The state of a widget is made of ...

File: crates/api/src/widget/state.rs [hunk 2]
```diff
@@ -7,23 +7,80 @@ pub trait AsAny: Any {
...
+/// widget!(MyWidget<MyState>)
+/// ```
+/// 
+/// [`widget`]: ./trait.Widget.html
+/// [`AsAny`]: ./trait.AsAny.html
+
 pub trait State: AsAny {
-    /// Init is used for initial setup.
-    fn init(&mut self, _: &mut Registry, _: &mut Context) {}
+    /// Init is used for setup the initial state of a widget, setting fields to starting values, registering service(s).
```
Comment (tzargoth) [inline crates/api/src/widget/state.rs:52]: ... for setting up ... values, and registering ...

File: crates/api/src/widget/state.rs [hunk 3]
```diff
@@ -7,23 +7,80 @@ pub trait AsAny: Any {
...
+    /// # Arguments
+    /// * `_registry`: Gives you access to the global Service Registry.
+    /// * `_ctx`: Represents the context of the current widget.Lets you manipulate the widget tree.
+    fn init(&mut self, _registry: &mut Registry, _ctx: &mut Context) {}
 
     /// Used to cleanup the state and is called after window close is requested.
-    fn cleanup(&mut self, _: &mut Registry, _: &mut Context) {}
+    /// # Arguments
+    /// * `_registry`: Gives you access to the global Service Registry.
+    /// * `_ctx`: Represents the context of the current widget.Lets you manipulate the widget tree.
```
Comment (tzargoth) [inline crates/api/src/widget/state.rs:63]: ... widget. Allows manipulation of ...

File: crates/api/src/widget/state.rs [hunk 4]
```diff
@@ -7,23 +7,80 @@ pub trait AsAny: Any {
...
+/// [`AsAny`]: ./trait.AsAny.html
+
 pub trait State: AsAny {
-    /// Init is used for initial setup.
-    fn init(&mut self, _: &mut Registry, _: &mut Context) {}
+    /// Init is used for setup the initial state of a widget, setting fields to starting values, registering service(s).
+    /// It is called after the widget is created.
+    /// 
+    /// # Arguments
+    /// * `_registry`: Gives you access to the global Service Registry.
```
Comment (tzargoth) [inline crates/api/src/widget/state.rs:56]: I don't think using "you" is appropriate for documentation.  Should be perspective agnostic.  So:
Provides access ...

File: crates/api/src/widget/state.rs [hunk 5]
```diff
@@ -7,23 +7,80 @@ pub trait AsAny: Any {
...
+    /// * `_ctx`: Represents the context of the current widget.Lets you manipulate the widget tree.
```
Comment (tzargoth) [inline crates/api/src/widget/state.rs:57]: ... widget.  Allows manipulation ...

File: crates/api/src/widget/state.rs [hunk 6]
```diff
@@ -7,23 +7,80 @@ pub trait AsAny: Any {
...
+    fn init(&mut self, _registry: &mut Registry, _ctx: &mut Context) {}
 
     /// Used to cleanup the state and is called after window close is requested.
-    fn cleanup(&mut self, _: &mut Registry, _: &mut Context) {}
+    /// # Arguments
+    /// * `_registry`: Gives you access to the global Service Registry.
```
Comment (tzargoth) [inline crates/api/src/widget/state.rs:62]: Provides access ...

File: crates/api/src/widget/state.rs [hunk 7]
```diff
@@ -7,23 +7,80 @@ pub trait AsAny: Any {
...
 
-    /// Updates the state for the given `ctx`.
-    ///
-    /// This update method is called before layout is calculated.
-    fn update(&mut self, _: &mut Registry, _: &mut Context) {}
+    /// Updates the state of a widget **before layout is calculated** for the given context when the widget becomes "dirty",
+    /// (e.g.: a property of a widget is changed or an [`event`] is fired)
+    /// 
+    /// # Arguments
+    /// * `_registry`: Gives you access to the global Service Registry.
```
Comment (tzargoth) [inline crates/api/src/widget/state.rs:70]: Provides access ...

File: crates/api/src/widget/state.rs [hunk 8]
```diff
@@ -7,23 +7,80 @@ pub trait AsAny: Any {
...
+    /// * `_ctx`: Represents the context of the current widget.Lets you manipulate the widget tree.
```
Comment (tzargoth) [inline crates/api/src/widget/state.rs:71]: ... widget.  Allows manipulation of ...

File: crates/api/src/widget/state.rs [hunk 9]
```diff
@@ -7,23 +7,80 @@ pub trait AsAny: Any {
...
+    /// [`event`]: ../trait.Event.html
+    fn update(&mut self, _registry: &mut Registry, _ctx: &mut Context) {}
 
-    /// Updates the state for the given `ctx`.
+    /// Updates the state **after layout is calculated and before rendering**
+    /// for the given context when the widget becomes "dirty",
+    /// (e.g.: a property of a widget is changed, or an [`event`] is fired)
+    ///
+    /// # Arguments
+    /// * `_registry`: Gives you access to the global Service Registry.
```
Comment (tzargoth) [inline crates/api/src/widget/state.rs:81]: Provides access ...

File: crates/api/src/widget/state.rs [hunk 10]
```diff
@@ -7,23 +7,80 @@ pub trait AsAny: Any {
...
+    /// * `_ctx`: Represents the context of the current widget.Lets you manipulate the widget tree.
```
Comment (tzargoth) [inline crates/api/src/widget/state.rs:82]: widget. Allows manipulation of ...

File: crates/api/src/widget/template.rs [hunk 11]
```diff
@@ -4,17 +4,62 @@ use crate::prelude::*;
...
 pub trait Template: Sized {
-    /// Creates the template of the widget and returns it.
+    /// Builds the template of the widget and returns it.
+    /// 
+    /// # Arguments
+    /// * `_id`: The id (Entity) of the instantiated widget in the Entity Store
+    /// * `_context`: The BuildContext used to build and instantiate new widgets
+    /// 
+    /// # Example
+    /// Lets define a widget called MyWidget with min,max and val properties with type of usize,
```
Comment (tzargoth) [inline crates/api/src/widget/template.rs:21]: Define a ... min, max ... usize, and then

**Comment by kivimango** [comment]
@tzargoth thank you.I have made the all the changes you suggested.

**Pull request merged**